File: ep08-dh-fk-matrix/cpp/dh_fk.cpp
#include <cmath>
#include <iomanip>
#include <iostream>

struct T4 { double m[4][4]; };

T4 dh(double a, double alpha, double d, double theta) {
    double ca=std::cos(alpha), sa=std::sin(alpha), ct=std::cos(theta), st=std::sin(theta);
    return {{{ct,-st*ca,st*sa,a*ct},{st,ct*ca,-ct*sa,a*st},{0,sa,ca,d},{0,0,0,1}}};
}

T4 mul(const T4& a, const T4& b) {
    T4 out{};
    for (int i=0;i<4;++i) for (int j=0;j<4;++j) for (int k=0;k<4;++k) out.m[i][j]+=a.m[i][k]*b.m[k][j];
    return out;
}

int main() {
    auto t=mul(dh(2,0,0,0.5236),dh(1.5,0,0,0.7854));
    std::cout<<std::fixed<<std::setprecision(4)<<"POSE "<<t.m[0][3]<<","<<t.m[1][3]<<","<<t.m[2][3]<<"\n";
    for (int angle=0;angle<=90;angle+=15) {
        double r=angle*0.01745329252;
        std::cout<<std::setw(2)<<std::setfill('0')<<angle<<" "<<2*std::cos(0.5235987756)+1.5*std::cos(0.5235987756+r)<<","<<2*std::sin(0.5235987756)+1.5*std::sin(0.5235987756+r)<<"\n";
    }
}

File: ep08-dh-fk-matrix/python/dh_fk.py
import csv
import math


def dh(a, alpha, d, theta):
    ca, sa = math.cos(alpha), math.sin(alpha)
    ct, st = math.cos(theta), math.sin(theta)
    return [[ct, -st * ca, st * sa, a * ct],
            [st, ct * ca, -ct * sa, a * st],
            [0.0, sa, ca, d],
            [0.0, 0.0, 0.0, 1.0]]


def mul(left, right):
    return [[sum(left[i][k] * right[k][j] for k in range(4)) for j in range(4)] for i in range(4)]


def fk_planar(theta2_deg):
    t1, t2 = math.radians(30.0), math.radians(theta2_deg)
    return 2.0 * math.cos(t1) + 1.5 * math.cos(t1 + t2), 2.0 * math.sin(t1) + 1.5 * math.sin(t1 + t2)


def sweep():
    return [(angle, *fk_planar(angle)) for angle in range(0, 91, 15)]


def write_csv(rows, path="traj.csv"):
    with open(path, "w", newline="", encoding="utf-8") as stream:
        writer = csv.writer(stream)
        writer.writerow(["theta2_deg", "x", "y"])
        writer.writerows((angle, f"{x:.4f}", f"{y:.4f}") for angle, x, y in rows)


if __name__ == "__main__":
    matrix = mul(dh(2.0, 0.0, 0.0, 0.5236), dh(1.5, 0.0, 0.0, 0.7854))
    print(f"POSE {matrix[0][3]:.4f},{matrix[1][3]:.4f},{matrix[2][3]:.4f}")
    rows = sweep()
    for angle, x, y in rows:
        print(f"{angle:02d} {x:.4f},{y:.4f}")
    write_csv(rows)
